addons/mrp/report/mrp_report_bom_structure.py

```diff
@@ -122,6 +122,41 @@ def _get_report_data(self, bom_id, searchQty=0, searchVariant=False):
             'precision': self.env['decimal.precision'].precision_get('Product Unit of Measure'),
         }
 
+    @api.model
+    def _get_components_closest_forecasted(self, lines, line_quantities, parent_bom, product_info, ignore_stock=False):
+        """
+            Returns a dict mapping a product to the closest date in the forecast report where
+            consummed quantity >= forecasted quantity.
+        """
+        if ignore_stock:
+            return {}
+        closest_forecasted = {}
+        remaining_products = self.env['product.product']
+        product_quantities_info = {}
+        for line in lines:
+            product = line.product_id
+            quantities_info = product_quantities_info.setdefault(product.id, self._get_quantities_info(product, line.product_uom_id, parent_bom, product_info))
+            stock_loc = quantities_info['stock_loc']
+            product_info[product.id]['consumptions'][stock_loc] += line_quantities.get(line.id, 0.0)
+            if product.detailed_type != 'product' or quantities_info['stock_loc'] != 'in_stock':
+                # Use date.min as a sentinel value for _get_stock_availability
+                closest_forecasted[product.id] = date.min
+            elif float_compare(product_info[product.id]['consumptions'][stock_loc], quantities_info['free_qty'], precision_rounding=product.uom_id.rounding) <= 0:
+                closest_forecasted[product.id] = date.min
+            else:
+                remaining_products |= product
+                closest_forecasted[product.id] = date.max
+        date_today = self.env.context.get('from_date', fields.date.today())
+        domain = [('state', '=', 'forecast'), ('date', '>=', date_today), ('product_id', 'in', remaining_products.ids)]
+        if self.env.context.get('warehouse'):
+            domain.append(('warehouse_id', '=', self.env.context.get('warehouse')))
+        res = self.env['report.stock.quantity']._read_group(domain, ['min_date:min(date)', 'product_id', 'product_qty'], ['product_id', 'product_qty'])
+        for group in res:
+            product_id = group['product_id'][0]
+            if group['product_qty'] >= product_info[product_id]['consumptions'][product_quantities_info[product_id]['stock_loc']]:
+                closest_forecasted[product_id] = min(closest_forecasted[product_id], group['min_date'])
```

Comment:
Can `closest_forecasted['product_id']` be something else than `date.max` at this point? If not, not sure we need that `min()`.

Reply:
I initially forgot to add `lazy=False` to the `_read_group`call but now it should be possible to have multiple `product_qty`for the same `product_id`, meaning that the `min()`call should be required.

Reply:
Then you'll have a traceback when you're in that case, since you'll be comparing `list` (from `closest_forecasted[product_id]`) and `date` (from `group['min_date']`).

Reply:
I forgot to change that line now that `closest_forecasted[product_id]` is a list :disappointed: